{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "stock-maple",
   "metadata": {},
   "source": [
    "#### CTS stability 2\n",
    "- In this version I will get out the genes which have only symbol, no FBg and see if they can be converted to ID with Flybase\n",
    "- Plot stability of RNAs enriched in specific cell types.\n",
    "- Use the data from L3 brain single-cell sequencing from the Cocanaugher 2022 paper\n",
    "- Show the fraction of TF, RBP, etc. in this dataset. Unfortunately we cannot do GO analysis on these genes since we do not know the background set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "current-austin",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Imports\n",
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import math\n",
    "import gffutils\n",
    "import scipy.stats as stats\n",
    "from collections import defaultdict\n",
    "\n",
    "sys.path.append('../scripts')\n",
    "from plot_helpers import *\n",
    "from plotting_fxns import PrettyBox\n",
    "from utilities import load_dataset\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "streaming-heritage",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the stability data\n",
    "outdir = '../Figures/CTS'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "#Load stability data\n",
    "rate_df = load_dataset('../Figures/summary_files/INSPEcT_rates.csv', '../Figures/summary_files/brain4sU_passed.csv')\n",
    "\n",
    "#Use the Flybase geneset for transcription factor\n",
    "TF_file = '../Figures/genesets/all_TFs.csv'\n",
    "tfs = set(pd.read_csv(TF_file, header=None)[0].values)\n",
    "\n",
    "#Import the GO slim based TF and RBP IDs:\n",
    "RBPs_go_file = '../Figures/GO/ens/RNA binding.txt'\n",
    "mRBPs_go_file = '../Figures/GO/ens/mRNA binding.txt'\n",
    "gial_file = '../../resources/glial_studies/glia-protrusion-localised-id-interest.txt'\n",
    "rbps = set(pd.read_csv(RBPs_go_file, header=None)[0].values)\n",
    "mrbps = set(pd.read_csv(mRBPs_go_file, header=None)[0].values)\n",
    "glial = set(pd.read_csv(gial_file, sep='\\t')['dmel_gene_id'].values)\n",
    "rbps_only = rbps.difference(mrbps)\n",
    "\n",
    "rate_df['TF'] = rate_df.index.isin(tfs)\n",
    "rate_df['RBP_all'] = rate_df.index.isin(rbps)\n",
    "rate_df['mRBP'] = rate_df.index.isin(mrbps)\n",
    "rate_df['RBP'] = rate_df.index.isin(rbps_only)\n",
    "rate_df['glial'] = rate_df.index.isin(glial)\n",
    "gene_groups = ['TF', 'RBP', 'mRBP', 'glial']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "innovative-montreal",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Number in each category\n",
    "print('num tfs', len(tfs))\n",
    "print('num rbps', len(rbps))\n",
    "print('num mrbps', len(mrbps))\n",
    "#mrbps is a complete subset of rbps. Also make category of rbps which are not mrbps\n",
    "print('num rbps_only', len(rbps_only))\n",
    "print('num glial', len(glial))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "conditional-insertion",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load the genes enriched from single-cell sequencing in L3 brains\n",
    "sc_file = '../../resources/other_studies/Corrales_Supplementary/Supplementary_spreadsheet_2_Ncells_and_gene_markers_per_cluster.xlsx'\n",
    "xl = pd.ExcelFile(sc_file)\n",
    "sheets = xl.sheet_names\n",
    "sheets.remove('Ncells-cluster')\n",
    "\n",
    "marker_dict = {}\n",
    "#replace some names because of typos in the spreadsheet\n",
    "celltype_replace = {'34 Immat': '34 Immat N'}\n",
    "for i in range(0, len(sheets)):\n",
    "# for i in range(0, 1):\n",
    "    df = pd.read_excel(sc_file, sheets[i])\n",
    "    ## celltype = ' '.join(sheets[0].split(' ')[1:])\n",
    "    celltype = sheets[i].strip()\n",
    "    if celltype in celltype_replace:\n",
    "        celltype = celltype_replace[celltype]\n",
    "    # Why don't the non-coding RNAs have an FBgn_ID? They do have them???\n",
    "    # For now use permissive cutoff where we will accept all the genes that they have identified as markers\n",
    "    pval_co = 1\n",
    "    fc_co = 1\n",
    "    #Currently, dropna() gets rid of lncRNAs, since they haven't been assigned to FBg IDs\n",
    "    genes = df.loc[(df['p_val_adj'] < pval_co) & (df['avg_log2FC'] > fc_co), ['gene', 'FBgn_ID']]\n",
    "\n",
    "    marker_dict[celltype] = genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d63f7a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "allgenes = pd.concat([*marker_dict.values()])\n",
    "no_id = allgenes[pd.isnull(allgenes['FBgn_ID'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d2fa76e",
   "metadata": {},
   "outputs": [],
   "source": [
    "no_id['gene'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c12001d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_dict['0 Immat N']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "316c9d80",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(marker_dict['0 Immat N'], columns=['gene_nam']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e5552ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_dict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "suffering-hazard",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all cell type markers and cell-type-specific TFs\n",
    "db = gffutils.FeatureDB(gffutils_db)\n",
    "genes = db.features_of_type('gene')\n",
    "ngenes = sum(1 for _ in genes)\n",
    "cts = set().union(*marker_dict.values())\n",
    "cts_tfs = cts.intersection(tfs)\n",
    "\n",
    "#Would be interesting to represent this visually somehow?\n",
    "print('num CTS RNAs', len(cts))\n",
    "print('num CTS TFs', len(cts_tfs))\n",
    "print('num TFs', len(tfs))\n",
    "print('frac CTS_TF/TF', len(cts_tfs)/len(tfs))\n",
    "print('frac TF/CTS', len(cts_tfs)/len(cts))\n",
    "print('frac CTS/genome', len(cts)/ngenes)\n",
    "print('frac TF/genome', len(tfs)/ngenes)\n",
    "\n",
    "# Write CTS genes to a file\n",
    "pd.Series(list(cts)).to_csv(os.path.join(outdir, 'CTS_genes.csv'), index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "greek-termination",
   "metadata": {},
   "source": [
    "#### Summarize stability percentile of cell-type-specific RNAs\n",
    "##### 1) Use all the clusters from Cocanougher (rate_df, clean_names)\n",
    "##### 2) Combine cell types of the same name to form general cell type (rate_df2, clean_names2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "comic-pulse",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine cell types of the same name to form a general cell type, e.g. Immat N\n",
    "def is_int(string):\n",
    "    try:\n",
    "        return int(string)\n",
    "    except:\n",
    "        return None\n",
    "\n",
    "def clean_name(string):\n",
    "    return ' '.join(string.split(' ')[1:])\n",
    "\n",
    "ct_names = list(marker_dict.keys())\n",
    "clean_names = [i for i in ct_names if is_int(i) is None]\n",
    "\n",
    "#Summarize by overall cell types\n",
    "rate_df2 = rate_df[['stab_percentile'] + gene_groups].copy()\n",
    "\n",
    "celltypes = defaultdict(set)\n",
    "for i in clean_names:\n",
    "    name = clean_name(i)\n",
    "    celltypes[name].add(i)\n",
    "\n",
    "#combine categories further as needed\n",
    "#e.g. Hemocytes will be labelled as Hemos and combined with category Hemos\n",
    "ct_combine = {'Hemocytes':'Hemos'}\n",
    "celltypes2 = {}\n",
    "for i in celltypes:\n",
    "    if i in ct_combine:\n",
    "        celltypes2[ct_combine[i]] = celltypes2[ct_combine[i]].union(celltypes[i])\n",
    "    else:\n",
    "        celltypes2[i] = celltypes[i]\n",
    "        \n",
    "#Assign genes to sub types to general cell type\n",
    "marker_dict2 = defaultdict(set)\n",
    "for i in celltypes2:\n",
    "    for j in celltypes2[i]:\n",
    "        marker_dict2[i] = marker_dict2[i].union(marker_dict[j])\n",
    "\n",
    "for ct in marker_dict2:\n",
    "    rate_df2[ct] = rate_df2.index.isin(marker_dict2[ct])\n",
    "\n",
    "ct_names2 = list(marker_dict2.keys())\n",
    "clean_names2 = [i for i in ct_names2 if is_int(i) is None]\n",
    "\n",
    "print('genes found specifically in at least one cell type', rate_df2[clean_names2].any(axis=1).value_counts())\n",
    "\n",
    "# In order to allow for genes to overlap cell types, we need to duplicate the gene index\n",
    "d = defaultdict(list)\n",
    "for cat in marker_dict2:\n",
    "    for gene in marker_dict2[cat]:\n",
    "        d[gene].append(cat)\n",
    "\n",
    "# Need to do this instead of melt inorder to allow genes to be a member of more than one category\n",
    "# https://stackoverflow.com/questions/42869544/dictionary-of-lists-to-dataframe\n",
    "cts_df = pd.DataFrame(pd.concat({k: pd.Series(v) for k, v in d.items()}).droplevel(1), columns=['celltype'])\n",
    "big_df = pd.merge(rate_df2[['stab_percentile'] + gene_groups], cts_df, left_on='gene', right_index=True, how='left').dropna(subset=['celltype'])\n",
    "#convert to longform df\n",
    "# big_df = rate_df2.reset_index().melt(['gene', 'stab_percentile', 'TF', 'RBP', 'mRBP'], var_name='celltype').query('value == True').drop(labels='value', axis=1)\n",
    "\n",
    "#https://stackoverflow.com/questions/41709239/dividing-dataframe-column-by-matching-index-in-another-dataframe\n",
    "cat_counts = big_df.groupby('celltype')[gene_groups].sum()\n",
    "cat_total = big_df.groupby('celltype')[gene_groups].size()\n",
    "cat_counts['total'] = cat_total\n",
    "for i in gene_groups:\n",
    "    cat_counts['%s_frac' % i] = cat_counts[i]/cat_counts['total']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "timely-regression",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Version of plot with RBPs and mRBPs included\n",
    "#https://stackoverflow.com/questions/47585775/seaborn-heatmap-with-single-column\n",
    "#https://stackoverflow.com/questions/33158075/custom-annotation-seaborn-heatmap\n",
    "\n",
    "# fig = plt.figure(figsize=(dfig*1.5, dfig*1.35), constrained_layout=True)\n",
    "fig = plt.figure(figsize=(dfig*4, dfig*4), constrained_layout=True)\n",
    "\n",
    "gs = fig.add_gridspec(ncols = 20, nrows = 1)\n",
    "gs.update(wspace=20) # set the spacing between axes. \n",
    "#0:10, and 10: did work, but 1:11 now doesn't work?\n",
    "ax = fig.add_subplot(gs[:, 0:10])\n",
    "heatmap_ax = fig.add_subplot(gs[:, 10:])\n",
    "order = big_df.groupby('celltype')['stab_percentile'].median().sort_values().index\n",
    "#fig.set_constrained_layout_pads(w_pad=0.2)\n",
    "ax = PrettyBox(data=big_df, x='stab_percentile', y='celltype', order=order, orient='h', color=color_dict['grey'], ax = ax)\n",
    "ax.set_ylabel('cell type (num genes)')\n",
    "# add the gene numbers to the cell type genes\n",
    "for l in ax.get_yticklabels():\n",
    "    ngenes = cat_counts.loc[l.get_text(), 'total']\n",
    "    l.set_text('%s (%s)' % (l.get_text(), ngenes))\n",
    "ax.set_yticklabels(ax.get_yticklabels())\n",
    "#txt = ax.text(-2, 0.5, 'cell type (num genes)', ha='center', va='center', rotation=90, transform=ax.transAxes)\n",
    "#_ = ax.set_yticklabels(labels[order])\n",
    "_ = ax.set_xlabel('stability percentile')\n",
    "heatmap_ax = sns.heatmap(cat_counts.loc[order][['TF_frac', 'RBP_frac', 'mRBP_frac']],\n",
    "                         annot=cat_counts.loc[order][['TF', 'RBP', 'mRBP']],\n",
    "                         ax = heatmap_ax, cmap='viridis',cbar_kws={'label':'fraction of genes', 'fraction':0.15,\n",
    "                                                                   'shrink':0.4, 'pad':0.02, 'aspect':15})\n",
    "heatmap_ax.set_yticks([])\n",
    "heatmap_ax.set_ylabel('')\n",
    "heatmap_ax.set_xticklabels(['TFs', 'RBPs', 'mRBPs'])\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'cts_box'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "controlled-positive",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Version of plot with TFs only. Try plotting without the gridspec and constrained_layout because it is causing me problems\n",
    "#https://stackoverflow.com/questions/47585775/seaborn-heatmap-with-single-column\n",
    "#https://stackoverflow.com/questions/33158075/custom-annotation-seaborn-heatmap\n",
    "fig = plt.figure(figsize=(dfig*1.5, dfig*1.35))\n",
    "ax = fig.add_axes((0.35, 0.15, 0.35, 0.8))\n",
    "heatmap_ax = fig.add_axes((0.7, 0.15, 0.1, 0.8))\n",
    "    \n",
    "ax = PrettyBox(data=big_df, x='stab_percentile', y='celltype', order=order, orient='h', color=color_dict['grey'], ax = ax)\n",
    "_ = ax.set_xlabel('stability percentile')\n",
    "_ = ax.set_ylabel('cell type (num genes)')\n",
    "\n",
    "# add the gene numbers to the cell type genes\n",
    "for l in ax.get_yticklabels():\n",
    "    ngenes = cat_counts.loc[l.get_text(), 'total']\n",
    "    l.set_text('%s (%s)' % (l.get_text(), ngenes))\n",
    "ax.set_yticklabels(ax.get_yticklabels())\n",
    "\n",
    "heatmap_ax = sns.heatmap(cat_counts.loc[order][['TF_frac']],\n",
    "                         annot=cat_counts.loc[order][['TF']],\n",
    "                         ax = heatmap_ax, cmap='viridis', square=True, cbar_kws={'label':'fraction of genes', 'fraction':0.15,\n",
    "                                                                   'shrink':0.4, 'pad':0.1, 'aspect':15})\n",
    "heatmap_ax.set_yticks([])\n",
    "heatmap_ax.set_xticklabels(['TFs'])\n",
    "heatmap_ax.set_ylabel('')\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'cts_box2'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8763dcf-95f3-4bff-96ed-a75cd179f040",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Version of plot with the TFs overlaid in swarmplot form\n",
    "fig = plt.figure(figsize=(dfig*1.5, dfig*1.35))\n",
    "ax = fig.add_axes((0.35, 0.15, 0.35, 0.8))\n",
    "heatmap_ax = fig.add_axes((0.7, 0.15, 0.1, 0.8))\n",
    "\n",
    "ax = sns.swarmplot(data=big_df, x='stab_percentile', y='celltype', order=order, orient='h', hue='TF', hue_order=[False, True], palette=[color_dict['grey'], color_dict['blue']], s=1, ax=ax)\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'cts_box3'), out_fmt), dpi = out_dpi)\n",
    "ax.text(0.5, 1, 'TF RNAs', transform=ax.transAxes, color=color_dict['blue'], ha='center')\n",
    "ax.get_legend().remove()\n",
    "\n",
    "_ = ax.set_xlabel('stability percentile')\n",
    "_ = ax.set_ylabel('cell type (num genes)')\n",
    "\n",
    "# add the gene numbers to the cell type genes\n",
    "for l in ax.get_yticklabels():\n",
    "    ngenes = cat_counts.loc[l.get_text(), 'total']\n",
    "    l.set_text('%s (%s)' % (l.get_text(), ngenes))\n",
    "ax.set_yticklabels(ax.get_yticklabels())\n",
    "\n",
    "heatmap_ax = sns.heatmap(cat_counts.loc[order][['TF_frac']],\n",
    "                         annot=cat_counts.loc[order][['TF']],\n",
    "                         ax = heatmap_ax, cmap='viridis', square=True, cbar_kws={'label':'fraction of genes', 'fraction':0.15,\n",
    "                                                                   'shrink':0.4, 'pad':0.1, 'aspect':15})\n",
    "heatmap_ax.set_yticks([])\n",
    "heatmap_ax.set_xticklabels(['TFs'])\n",
    "heatmap_ax.set_ylabel('')\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'cts_box3'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5817ccb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Version of plot with the glial enriched genes overlaid in swarmplot form\n",
    "fig = plt.figure(figsize=(dfig*1.5, dfig*1.35))\n",
    "ax = fig.add_axes((0.35, 0.15, 0.35, 0.8))\n",
    "heatmap_ax = fig.add_axes((0.7, 0.15, 0.1, 0.8))\n",
    "\n",
    "ax = sns.swarmplot(data=big_df, x='stab_percentile', y='celltype', order=order, orient='h', hue='glial', hue_order=[False, True], palette=[color_dict['grey'], color_dict['blue']], s=1, ax=ax)\n",
    "ax.text(0.5, 1, 'glial protrusion RNAs', transform=ax.transAxes, color=color_dict['blue'], ha='center')\n",
    "ax.get_legend().remove()\n",
    "\n",
    "_ = ax.set_xlabel('stability percentile')\n",
    "_ = ax.set_ylabel('cell type (num genes)')\n",
    "\n",
    "# add the gene numbers to the cell type genes\n",
    "for l in ax.get_yticklabels():\n",
    "    ngenes = cat_counts.loc[l.get_text(), 'total']\n",
    "    l.set_text('%s (%s)' % (l.get_text(), ngenes))\n",
    "ax.set_yticklabels(ax.get_yticklabels())\n",
    "\n",
    "heatmap_ax = sns.heatmap(glial_df.loc[order][['frac']],\n",
    "                         annot=glial_df.loc[order][['total']],\n",
    "                         ax = heatmap_ax, cmap='viridis', square=True, cbar_kws={'label':'fraction of genes', 'fraction':0.15,\n",
    "                                                                   'shrink':0.4, 'pad':0.1, 'aspect':15})\n",
    "heatmap_ax.set_yticks([])\n",
    "heatmap_ax.set_xticklabels(['  count'])\n",
    "heatmap_ax.set_ylabel('')\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'cts_box_glia'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "wireless-resource",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We cannot do a proper GO analysis on this data since it is unclear what is the background set. \n",
    "#However, we could match these to the GO_slim_categories to see what they are enriched for as a rough idea.\n",
    "low_stab = big_df[big_df['stab_percentile'] < 25].copy()\n",
    "hi_stab = big_df[big_df['stab_percentile'] > 75].copy()\n",
    "\n",
    "slim_dict = {}\n",
    "indir = '../Figures/GO/ens/'\n",
    "for file in os.listdir(indir):\n",
    "    slim_dict[file.rstrip('.txt')] = set(pd.read_csv(os.path.join(indir, file), header=None)[0].tolist())\n",
    "    \n",
    "d = {'low':{}, 'hi':{}}\n",
    "hi_genes = set(hi_stab['gene'].tolist())\n",
    "low_genes = set(low_stab['gene'].tolist())\n",
    "n_low = len(low_genes)\n",
    "n_hi = len(hi_genes)\n",
    "for i in slim_dict:\n",
    "    d['low'][i] = len(low_genes.intersection(slim_dict[i]))/n_low\n",
    "    d['hi'][i] = len(hi_genes.intersection(slim_dict[i]))/n_hi\n",
    "pd.DataFrame.from_dict(d, orient='columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "compact-franklin",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot ridgeline plot\n",
    "cats = order\n",
    "num_cats = len(cats)\n",
    "pal = sns.cubehelix_palette(num_cats, rot=-.25, light=.7)\n",
    "#sns.set_palette(pal)\n",
    "fig = plt.figure(figsize=(dfig, dfig*2))\n",
    "gs = fig.add_gridspec(ncols=1, nrows = len(cats))\n",
    "for i in range(len(cats)):\n",
    "    fig.add_subplot(gs[i])\n",
    "    ax = sns.kdeplot(x='stab_percentile', data=big_df[big_df['celltype']==cats[i]], bw_adjust=.5, clip_on=False,\n",
    "      fill=True, alpha=1, linewidth=1.5, color=pal[i])\n",
    "    ax = sns.kdeplot(x='stab_percentile', data=big_df[big_df['celltype']==cats[i]], clip_on=False, color=\"w\", lw=2, bw_adjust=.5)\n",
    "    ax.spines['left'].set_visible(False)\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax.set_fc((0.0, 0.0, 0.0, 0.0))\n",
    "    ax.set_yticks([])\n",
    "    ax.set_ylabel('')\n",
    "    if i != len(cats)-1:\n",
    "        ax.set_xticks([])\n",
    "        ax.set_xlabel('')\n",
    "    else:\n",
    "        ax.set_xlabel('stability percentile')\n",
    "    label = cats[i]\n",
    "#     if cats[i] in cat_pretty_names:\n",
    "#         label = cat_pretty_names[cats[i]]\n",
    "#     else:\n",
    "#         label = cats[i]\n",
    "    ax.text(-0.25, .3, label, fontsize=5, fontweight=\"bold\",\n",
    "            ha=\"left\", va=\"center\", transform=ax.transAxes, color=pal[i])\n",
    "\n",
    "fig.subplots_adjust(hspace=-.3)\n",
    "#this adds space on the left side to make room for the labels\n",
    "fig.subplots_adjust(left=0.2)\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'test_ridge'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "varying-strip",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot the stability percentile of TFs for specific cell types\n",
    "rate_df2['CTS'] = rate_df2[clean_names2].any(axis=1)\n",
    "fig = plt.figure(figsize=(dfig*0.7, dfig*1.35), constrained_layout=True)\n",
    "ax = PrettyBox(data=rate_df2, x='TF', y='stab_percentile',  hue='CTS', fliersize = 0, width = 0.4)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(['non-TF', 'TF'])\n",
    "#move the ylabel in to get more room\n",
    "ax.set_ylabel('RNA stability percentile', labelpad=0)\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "#ax.legend(handles, ['distributed', 'cell type specific'], bbox_to_anchor=(0.5, 1.0), loc='lower center', title='RNA class')\n",
    "#prevent legend box from turning black.\n",
    "ax.legend().get_frame().set_fc((1.0,1.0,1.0,0.8))\n",
    "ax.legend(handles, ['distributed', 'cell type specific'], bbox_to_anchor=(0.5, 1.0), loc='lower center', title='RNA class')\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'stab_cts_TFs'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "scheduled-omaha",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Overall percentile median = 50, as expected\n",
    "print('stability of all genes %s' % rate_df2['stab_percentile'].median())\n",
    "#Percentile of CTS genes\n",
    "print('stability of CTS genes %s' % rate_df2.loc[rate_df2['CTS'], 'stab_percentile'].median())\n",
    "#Percentile of non-CTS genes\n",
    "print('stability of non-CTS genes %s' % rate_df2.loc[~rate_df2['CTS'], 'stab_percentile'].median())\n",
    "#Percentile of TF genes\n",
    "print('stability of TF genes %s' % rate_df2.loc[rate_df2['TF'], 'stab_percentile'].median())\n",
    "#Percentile of non-CTS TF genes\n",
    "print('stability of non-CTS TF genes %s' % rate_df2.loc[rate_df2['TF'] & ~rate_df2['CTS'], 'stab_percentile'].median())\n",
    "#Percentile of TF & CTS\n",
    "print('stability of CTS & TF genes %s' % rate_df2.loc[rate_df2['TF'] & rate_df2['CTS'], 'stab_percentile'].median())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bacc0ef-e37b-4ceb-bd8e-82a756f932a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find CTS genes which are in the top and bottom 10% of most stable RNAs for the CTS genes\n",
    "# Previously I tried the top and bottom 10% of stability overall, but this found only 26 genes in the bottom 10% of genes\n",
    "# Use as bg set, all the genes with stability measurements\n",
    "bg_genes = rate_df2.index\n",
    "\n",
    "geneset_outdir = os.path.join(outdir, 'genesets')\n",
    "os.makedirs(geneset_outdir, exist_ok=True)\n",
    "\n",
    "num_genes = int(round(len(rate_df2.query('CTS'))/10))\n",
    "sorted_cts = rate_df2.query('CTS').sort_values(by='stab_percentile')\n",
    "stab_genes10 = sorted_cts.head(n=num_genes).index\n",
    "unstab_genes10 = sorted_cts.tail(n=num_genes).index\n",
    "\n",
    "# Write the output genelists\n",
    "pd.DataFrame(stab_genes10).to_csv(os.path.join(geneset_outdir, 'CTS_10unstable_genes.csv'), header=None, index=None)\n",
    "pd.DataFrame(unstab_genes10).to_csv(os.path.join(geneset_outdir, 'CTS_10stable_genes.csv'), header=None, index=None)\n",
    "pd.DataFrame(bg_genes).to_csv(os.path.join(geneset_outdir, 'bg_genes.csv'), header=None, index=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 ('pretty')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "bdb49103a4ed208a05ea4530afbe53462c06fafea10c7833b005d674746fdb08"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
