{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "stock-maple",
   "metadata": {},
   "source": [
    "#### CTS stability\n",
    "- Plot stability of RNAs enriched in specific cell types.\n",
    "- Use the data from L3 brain single-cell sequencing from the Cocanaugher 2022 paper\n",
    "- Show the fraction of TF, RBP, etc. in this dataset. Unfortunately we cannot do GO analysis on these genes since we do not know the background set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "current-austin",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Imports\n",
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import math\n",
    "import gffutils\n",
    "import scipy.stats as stats\n",
    "from collections import defaultdict\n",
    "import warnings\n",
    "\n",
    "sys.path.append('../scripts')\n",
    "from plot_helpers import *\n",
    "from plotting_fxns import PrettyBox\n",
    "from utilities import load_dataset\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "streaming-heritage",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the stability data\n",
    "outdir = '../Figures/CTS'\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "#Load stability data\n",
    "rate_df = load_dataset('../Figures/summary_files/INSPEcT_rates.csv', '../Figures/summary_files/brain4sU_passed.csv')\n",
    "\n",
    "#Use the Flybase geneset for transcription factor\n",
    "TF_file = '../Figures/genesets/all_TFs.csv'\n",
    "tfs = set(pd.read_csv(TF_file, header=None)[0].values)\n",
    "\n",
    "#Import the GO slim based TF and RBP IDs:\n",
    "RBPs_go_file = '../Figures/GO/ens/RNA binding.txt'\n",
    "mRBPs_go_file = '../Figures/GO/ens/mRNA binding.txt'\n",
    "gial_file = '../../resources/glial_studies/glia-protrusion-localised-id-interest.txt'\n",
    "rbps = set(pd.read_csv(RBPs_go_file, header=None)[0].values)\n",
    "mrbps = set(pd.read_csv(mRBPs_go_file, header=None)[0].values)\n",
    "glial = set(pd.read_csv(gial_file, sep='\\t')['dmel_gene_id'].values)\n",
    "rbps_only = rbps.difference(mrbps)\n",
    "\n",
    "rate_df['TF'] = rate_df.index.isin(tfs)\n",
    "rate_df['RBP_all'] = rate_df.index.isin(rbps)\n",
    "rate_df['mRBP'] = rate_df.index.isin(mrbps)\n",
    "rate_df['RBP'] = rate_df.index.isin(rbps_only)\n",
    "rate_df['glial'] = rate_df.index.isin(glial)\n",
    "gene_groups = ['TF', 'RBP', 'mRBP', 'glial']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "innovative-montreal",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Number in each category\n",
    "print('num tfs', len(tfs))\n",
    "print('num rbps', len(rbps))\n",
    "print('num mrbps', len(mrbps))\n",
    "#mrbps is a complete subset of rbps. Also make category of rbps which are not mrbps\n",
    "print('num rbps_only', len(rbps_only))\n",
    "print('num glial', len(glial))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d9a1cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the genes enriched from single-cell sequencing in L3 brains\n",
    "cts_df = pd.read_csv(os.path.join(outdir, 'cts_celltypes_log1.00.csv'), index_col=0)\n",
    "cts_df.index.name = 'gene'\n",
    "# Index cts_df2 on the cell type to get all genes associated with certain cell types\n",
    "cts_df2 = cts_df.reset_index().set_index('celltype').copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "suffering-hazard",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all cell type markers and cell-type-specific TFs\n",
    "db = gffutils.FeatureDB(gffutils_db)\n",
    "genes = db.features_of_type('gene')\n",
    "ngenes = sum(1 for _ in genes)\n",
    "cts = cts_df.index.unique()\n",
    "cts_tfs = cts.intersection(tfs)\n",
    "\n",
    "#Would be interesting to represent this visually somehow?\n",
    "print('num CTS RNAs', len(cts))\n",
    "print('num CTS TFs', len(cts_tfs))\n",
    "print('num TFs', len(tfs))\n",
    "print('frac CTS_TF/TF', len(cts_tfs)/len(tfs))\n",
    "print('frac TF/CTS', len(cts_tfs)/len(cts))\n",
    "print('frac CTS/genome', len(cts)/ngenes)\n",
    "print('frac TF/genome', len(tfs)/ngenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6accbd03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Summarize by overall cell types\n",
    "rate_df2 = rate_df[['deg_rate', 'stab_percentile'] + gene_groups].copy()\n",
    "\n",
    "cts_names = cts_df2.index.unique()\n",
    "for ct in cts_names:\n",
    "    # This causes an error if there's only one gene in it\n",
    "    rate_df2[ct] = rate_df2.index.isin(cts_df2.loc[ct, ['gene']].values.flatten())\n",
    "\n",
    "    \n",
    "print('genes found specifically in at least one cell type', rate_df2[cts_names].any(axis=1).value_counts())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a36802a",
   "metadata": {},
   "outputs": [],
   "source": [
    "big_df = pd.merge(rate_df2[['deg_rate', 'stab_percentile'] + gene_groups], cts_df, left_on='gene', right_index=True, how='left').dropna(subset=['celltype'])\n",
    "# https://stackoverflow.com/questions/41709239/dividing-dataframe-column-by-matching-index-in-another-dataframe\n",
    "cat_counts = big_df.groupby('celltype')[gene_groups].sum()\n",
    "cat_total = big_df.groupby('celltype')[gene_groups].size()\n",
    "cat_counts['total'] = cat_total\n",
    "for i in gene_groups:\n",
    "    cat_counts['%s_frac' % i] = cat_counts[i]/cat_counts['total']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "523ddbfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from plotting_fxns import sc_swarmplot, enrich_heatmap\n",
    "\n",
    "# Version of plot with the TFs overlaid in swarmplot form\n",
    "fig = plt.figure(figsize=(dfig*1.6, dfig*1.5))\n",
    "hstart = 0.22\n",
    "h = 0.95 - hstart\n",
    "ax = fig.add_axes((0.32, hstart, 0.29, h))\n",
    "\n",
    "order = big_df.groupby('celltype')['stab_percentile'].median().sort_values().index\n",
    "\n",
    "# Ingnore the seaborn swarm overplotting warning here:\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter('ignore')\n",
    "    ax = sc_swarmplot(data=big_df, all_genes=rate_df2, x='stab_percentile', y='celltype', hue='TF', hue_name='TF mRNAs', x_lab='stability percentile', \n",
    "                      y_lab='cell type (num genes)', ax=ax)\n",
    "                      \n",
    "res_tf = enrich_heatmap(data=big_df, all_genes=rate_df2, x='stab_percentile', y='celltype', hue=['TF'], y_lab1='fraction of genes',\n",
    "                   y_lab2='-log'r'$_{10}$'' p-value', hstart=hstart, lstart=0.65, fig=fig, xticklabs1=['counts'], xticklabs2=['enrichment'], hm_xlab=False)\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'cts_tf_swarm3'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b09c8d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Version of plot with the gial protrusion genes overlaid in swarmplot form -- for Ilan\n",
    "fig = plt.figure(figsize=(dfig*1.6, dfig*1.5))\n",
    "hstart = 0.22\n",
    "h = 0.95 - hstart\n",
    "ax = fig.add_axes((0.32, hstart, 0.29, h))\n",
    "\n",
    "order = big_df.groupby('celltype')['stab_percentile'].median().sort_values().index\n",
    "\n",
    "# Ingnore the seaborn swarm overplotting warning here:\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter('ignore')\n",
    "    ax = sc_swarmplot(data=big_df, all_genes=rate_df2, x='stab_percentile', y='celltype', hue='glial', hue_name='glial protrusion mRNAs', x_lab='stability percentile', \n",
    "                      y_lab='cell type (num genes)', ax=ax)\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'cts_glial_swarm2'), out_fmt), dpi = 600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "444fa83f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Version of plot with the RBPs and mRBPs overlaid in swarmplot form\n",
    "fig = plt.figure(figsize=(dfig*1.9, dfig*1.5))\n",
    "hstart = 0.22\n",
    "h = 0.95 - hstart\n",
    "ax = fig.add_axes((0.28, hstart, 0.29, h))\n",
    "\n",
    "order = big_df.groupby('celltype')['stab_percentile'].median().sort_values().index\n",
    "\n",
    "# Ingnore the seaborn swarm overplotting warning here:\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter('ignore')\n",
    "    ax = sc_swarmplot(data=big_df, all_genes=rate_df2, x='stab_percentile', y='celltype', hue='RBP', hue_name='RBPs', x_lab='stability percentile', \n",
    "                      y_lab='cell type (num genes)', ax=ax)\n",
    "                      \n",
    "res_rbp = enrich_heatmap(data=big_df, all_genes=rate_df2, x='stab_percentile', y='celltype', hue=['RBP', 'mRBP'], y_lab1='fraction of genes',\n",
    "               y_lab2='-log'r'$_{10}$'' p-value', hstart=hstart, lstart=0.61, fig=fig, xticklabs1=['RBP', 'mRBP'], xticklabs2=['RBP', 'mRBP'],\n",
    "               hm_width=0.09)\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'cts_rbp_swarm'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9814f6e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot enrichment of RBPs and mRBPs\n",
    "fig = plt.figure(figsize=(dfig*1.2, dfig*1.5))\n",
    "hstart = 0.15\n",
    "h = 0.95 - hstart\n",
    "\n",
    "# Plot heatmap of the counts and enrichment values\n",
    "arrays = enrich_heatmap(data=big_df, all_genes=rate_df2, x='stab_percentile', y='celltype', hue=['RBP', 'mRBP'], y_lab1='fraction of genes',\n",
    "                   y_lab2='-log'r'$_{10}$'' p-value', lstart=0.5, hstart=0.22, hm_width=0.1, cb_width=0.03, fig=fig, xticklabs1=['RBP', 'mRBP'],\n",
    "                   xticklabs2=['RBP', 'mRBP'], ylabels=True, cbar_lab_sp=6)\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'cts_rbps'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "758ed0af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Version of plot with the gial protrusion genes overlaid in swarmplot form -- for Ilan\n",
    "fig = plt.figure(figsize=(dfig*1.6, dfig*1.5))\n",
    "hstart = 0.22\n",
    "h = 0.95 - hstart\n",
    "ax = fig.add_axes((0.32, hstart, 0.29, h))\n",
    "\n",
    "order = big_df.groupby('celltype')['stab_percentile'].median().sort_values().index\n",
    "\n",
    "# Ingnore the seaborn swarm overplotting warning here:\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter('ignore')\n",
    "    ax = sc_swarmplot(data=big_df, all_genes=rate_df2, x='stab_percentile', y='celltype', hue='glial', hue_name='glial protrusion mRNAs', x_lab='stability percentile', \n",
    "                      y_lab='cell type (num genes)', ax=ax)\n",
    "                      \n",
    "enrich_heatmap(data=big_df, all_genes=rate_df2, x='stab_percentile', y='celltype', hue=['glial'], y_lab1='fraction of genes',\n",
    "                   y_lab2='-log'r'$_{10}$'' p-value', lstart=0.65, hstart=0.22, fig=fig, xticklabs1=['counts'], xticklabs2=['enrichment'], hm_xlab=False)\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'cts_glial_swarm'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "wireless-resource",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We cannot do a proper GO analysis on this data since it is unclear what is the background set. \n",
    "#However, we could match these to the GO_slim_categories to see what they are enriched for as a rough idea.\n",
    "low_stab = big_df[big_df['stab_percentile'] < 25].copy()\n",
    "hi_stab = big_df[big_df['stab_percentile'] > 75].copy()\n",
    "\n",
    "slim_dict = {}\n",
    "indir = '../Figures/GO/ens/'\n",
    "for file in os.listdir(indir):\n",
    "    slim_dict[file.rstrip('.txt')] = set(pd.read_csv(os.path.join(indir, file), header=None)[0].tolist())\n",
    "    \n",
    "d = {'low':{}, 'hi':{}}\n",
    "hi_genes = set(hi_stab.index.tolist())\n",
    "low_genes = set(low_stab.index.tolist())\n",
    "n_low = len(low_genes)\n",
    "n_hi = len(hi_genes)\n",
    "for i in slim_dict:\n",
    "    d['low'][i] = len(low_genes.intersection(slim_dict[i]))/n_low\n",
    "    d['hi'][i] = len(hi_genes.intersection(slim_dict[i]))/n_hi\n",
    "pd.DataFrame.from_dict(d, orient='columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8b8db55",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot the stability percentile of TFs for specific cell types\n",
    "celltypes = cts_df['celltype'].unique()\n",
    "rate_df2['CTS'] = rate_df2[celltypes].any(axis=1)\n",
    "\n",
    "from plotting_fxns import get_boxtop, add_stars, get_box_coords\n",
    "\n",
    "fig = plt.figure(figsize=(dfig*0.75, dfig*1.5), constrained_layout=True)\n",
    "ax = PrettyBox(data=rate_df2, x='TF', y='stab_percentile',  hue='CTS', fliersize = 0, width = 0.4)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(['non-TF', 'TF'])\n",
    "#move the ylabel in to get more room\n",
    "ax.set_ylabel('RNA stability percentile', labelpad=0)\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "#ax.legend(handles, ['distributed', 'cell type specific'], bbox_to_anchor=(0.5, 1.0), loc='lower center', title='RNA class')\n",
    "#prevent legend box from turning black.\n",
    "ax.legend().get_frame().set_fc((1.0,1.0,1.0,0.8))\n",
    "ax.legend(handles, ['ubiquitous', 'cell type specific'], bbox_to_anchor=(0.5, 1.15), loc='lower center', title='RNA class')\n",
    "\n",
    "# First comparison: non-CTS, TF vs. non-TF\n",
    "# Second comparison: CTS, TF vs. non-TF\n",
    "d1a = rate_df2.query('~CTS & ~TF')['deg_rate']\n",
    "d1b = rate_df2.query('~CTS & TF')['deg_rate']\n",
    "d2a = rate_df2.query('CTS & ~TF')['deg_rate']\n",
    "d2b = rate_df2.query('CTS & TF')['deg_rate']\n",
    "\n",
    "_, p1 = stats.mannwhitneyu(d1a, d1b)\n",
    "_, p2 = stats.mannwhitneyu(d2a, d2b)\n",
    "\n",
    "h1 = max(get_boxtop(rate_df2, col1='CTS', val1=False, col2='TF', val2=False, val_col='stab_percentile'),\n",
    "         get_boxtop(rate_df2, col1='CTS', val1=False, col2='TF', val2=True, val_col='stab_percentile'))\n",
    "\n",
    "h2 = max(get_boxtop(rate_df2, col1='CTS', val1=True, col2='TF', val2=False, val_col='stab_percentile'),\n",
    "         get_boxtop(rate_df2, col1='CTS', val1=True, col2='TF', val2=True, val_col='stab_percentile'))\n",
    "\n",
    "add_stars(0, 2, starty=h1+2, height=2.5, p=p1, ax=ax)\n",
    "add_stars(1, 3, starty=h2+10, height=2.5, p=p2, ax=ax)\n",
    "ax.set_ylim(-2, 100)\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'stab_cts_TFs'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ebac6b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If we want to plot RBP and mRBP separately with Seaborn swarmplot, need to duplicate the mRBP rows\n",
    "# copy the mRBP rows out\n",
    "mRBP_df = rate_df2.query('mRBP').copy()[['deg_rate', 'stab_percentile', 'mRBP']]\n",
    "rate_df3 = rate_df2.copy()[['deg_rate', 'stab_percentile', 'RBP']]\n",
    "rate_df3 = pd.concat([mRBP_df, rate_df3]).fillna(False)\n",
    "# https://stackoverflow.com/questions/67095325/how-to-transform-multiple-boolean-columns-to-one-column-with-column-headers-and\n",
    "rate_df3['category'] = rate_df3[['mRBP', 'RBP']].idxmax(1).where(rate_df3[['mRBP', 'RBP']].any(1)).fillna('other')\n",
    "rate_df3['CTS'] = rate_df3.index.isin(rate_df2.query('CTS').index)\n",
    "# Check that each gene belongs to only one category\n",
    "assert len(rate_df3.query('RBP')) + len(rate_df3.query('mRBP')) + len(rate_df3.query('~RBP & ~mRBP')) == len(rate_df3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11fcaa0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot the stability percentile of RBPs for specific cell types\n",
    "from plotting_fxns import get_boxtop, add_stars, get_box_coords\n",
    "\n",
    "fig = plt.figure(figsize=(dfig*1.5, dfig*1.5), constrained_layout=True)\n",
    "ax = PrettyBox(data=rate_df3, x='category', y='stab_percentile', order=['other', 'RBP', 'mRBP'], hue='CTS', fliersize = 0, width = 0.4)\n",
    "ax.set_xlabel('')\n",
    "# ax.set_xticklabels(['non-RBP', 'RBP'])\n",
    "#move the ylabel in to get more room\n",
    "ax.set_ylabel('RNA stability percentile', labelpad=0)\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "#ax.legend(handles, ['distributed', 'cell type specific'], bbox_to_anchor=(0.5, 1.0), loc='lower center', title='RNA class')\n",
    "#prevent legend box from turning black.\n",
    "\n",
    "ax.legend().get_frame().set_fc((1.0,1.0,1.0,0.8))\n",
    "# top gets clipped off, even with constrained_layout\n",
    "# ax.legend(handles, ['ubiquitous', 'cell type specific'], bbox_to_anchor=(0.5, 1.3), loc='lower center', title='RNA class')\n",
    "ax.legend(handles, ['ubiquitous', 'cell type specific'], bbox_to_anchor=(1, 0.5), loc='center left', title='RNA class')\n",
    "\n",
    "\n",
    "# Note that for the comparison between other and mRBPs -- the other group is also depleted of RBPs in the graph but for the stats, \n",
    "# we will compare all non-mRBPs to mRBPs\n",
    "\n",
    "# First comparison: non-CTS, RBP vs. non-RBP\n",
    "# Second comparison: non-CTS, mRBP vs. non-mRBP\n",
    "# Third comparison: CTS, RBP vs. non-RBP\n",
    "# Fourth comparison: CTS, mRBP vs. non-mRBP\n",
    "d1a = rate_df3.query('~CTS & ~RBP')['deg_rate']\n",
    "d1b = rate_df3.query('~CTS & RBP')['deg_rate']\n",
    "\n",
    "d2a = rate_df3.query('~CTS & ~mRBP')['deg_rate']\n",
    "d2b = rate_df3.query('~CTS & mRBP')['deg_rate']\n",
    "\n",
    "d3a = rate_df3.query('CTS & ~RBP')['deg_rate']\n",
    "d3b = rate_df3.query('CTS & RBP')['deg_rate']\n",
    "\n",
    "d4a = rate_df3.query('CTS & ~mRBP')['deg_rate']\n",
    "d4b = rate_df3.query('CTS & mRBP')['deg_rate']\n",
    "\n",
    "_, p1 = stats.mannwhitneyu(d1a, d1b)\n",
    "_, p2 = stats.mannwhitneyu(d2a, d2b)\n",
    "_, p3 = stats.mannwhitneyu(d3a, d3b)\n",
    "_, p4 = stats.mannwhitneyu(d4a, d4b)\n",
    "\n",
    "h1 = max(get_boxtop(rate_df3, col1='CTS', val1=False, col2='RBP', val2=False, val_col='stab_percentile'),\n",
    "         get_boxtop(rate_df3, col1='CTS', val1=False, col2='RBP', val2=True, val_col='stab_percentile'))\n",
    "\n",
    "h2 = max(get_boxtop(rate_df3, col1='CTS', val1=False, col2='mRBP', val2=False, val_col='stab_percentile'),\n",
    "         get_boxtop(rate_df3, col1='CTS', val1=False, col2='mRBP', val2=True, val_col='stab_percentile'))\n",
    "\n",
    "h3 = max(get_boxtop(rate_df3, col1='CTS', val1=True, col2='RBP', val2=False, val_col='stab_percentile'),\n",
    "         get_boxtop(rate_df3, col1='CTS', val1=True, col2='RBP', val2=True, val_col='stab_percentile'))\n",
    "\n",
    "h4 = max(get_boxtop(rate_df3, col1='CTS', val1=True, col2='mRBP', val2=False, val_col='stab_percentile'),\n",
    "         get_boxtop(rate_df3, col1='CTS', val1=True, col2='mRBP', val2=True, val_col='stab_percentile'))\n",
    "\n",
    "add_stars(0, 2, starty=h1+2, height=2.5, p=p1, ax=ax)\n",
    "add_stars(0, 4, starty=h2+10, height=2.5, p=p2, ax=ax)\n",
    "add_stars(1, 3, starty=h3+18, height=2.5, p=p3, ax=ax)\n",
    "add_stars(1, 5, starty=h4+26, height=2.5, p=p4, ax=ax)\n",
    "\n",
    "ax.set_ylim(-2, 100)\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'stab_cts_RBPs2'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f83a15eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many genes in the CTS mRBP category?\n",
    "len(rate_df3.query('CTS & mRBP')['deg_rate'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad39dd1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot the stability percentile of RBPs and mRBPs for specific cell types\n",
    "from plotting_fxns import get_boxtop, add_stars, get_box_coords\n",
    "\n",
    "fig = plt.figure(figsize=(dfig*0.75, dfig*1.5), constrained_layout=True)\n",
    "ax = PrettyBox(data=rate_df2, x='RBP', y='stab_percentile',  hue='CTS', fliersize = 0, width = 0.4)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(['non-RBP', 'RBP'])\n",
    "#move the ylabel in to get more room\n",
    "ax.set_ylabel('RNA stability percentile', labelpad=0)\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "#ax.legend(handles, ['distributed', 'cell type specific'], bbox_to_anchor=(0.5, 1.0), loc='lower center', title='RNA class')\n",
    "#prevent legend box from turning black.\n",
    "ax.legend().get_frame().set_fc((1.0,1.0,1.0,0.8))\n",
    "ax.legend(handles, ['ubiquitous', 'cell type specific'], bbox_to_anchor=(0.5, 1.15), loc='lower center', title='RNA class')\n",
    "\n",
    "# First comparison: non-CTS, TF vs. non-TF\n",
    "# Second comparison: CTS, TF vs. non-TF\n",
    "d1a = rate_df2.query('~CTS & ~RBP')['deg_rate']\n",
    "d1b = rate_df2.query('~CTS & RBP')['deg_rate']\n",
    "d2a = rate_df2.query('CTS & ~RBP')['deg_rate']\n",
    "d2b = rate_df2.query('CTS & RBP')['deg_rate']\n",
    "\n",
    "_, p1 = stats.mannwhitneyu(d1a, d1b)\n",
    "_, p2 = stats.mannwhitneyu(d2a, d2b)\n",
    "\n",
    "h1 = max(get_boxtop(rate_df2, col1='CTS', val1=False, col2='RBP', val2=False, val_col='stab_percentile'),\n",
    "         get_boxtop(rate_df2, col1='CTS', val1=False, col2='RBP', val2=True, val_col='stab_percentile'))\n",
    "\n",
    "h2 = max(get_boxtop(rate_df2, col1='CTS', val1=True, col2='RBP', val2=False, val_col='stab_percentile'),\n",
    "         get_boxtop(rate_df2, col1='CTS', val1=True, col2='RBP', val2=True, val_col='stab_percentile'))\n",
    "\n",
    "add_stars(0, 2, starty=h1+2, height=2.5, p=p1, ax=ax)\n",
    "add_stars(1, 3, starty=h2+10, height=2.5, p=p1, ax=ax)\n",
    "\n",
    "plt.savefig('%s.%s' % (os.path.join(outdir, 'stab_cts_RBPs'), out_fmt), dpi = out_dpi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "scheduled-omaha",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Overall percentile median = 50, as expected\n",
    "print('stability of all genes %s' % rate_df2['stab_percentile'].median())\n",
    "#Percentile of CTS genes\n",
    "print('stability of CTS genes %s' % rate_df2.loc[rate_df2['CTS'], 'stab_percentile'].median())\n",
    "#Percentile of non-CTS genes\n",
    "print('stability of non-CTS genes %s' % rate_df2.loc[~rate_df2['CTS'], 'stab_percentile'].median())\n",
    "#Percentile of TF genes\n",
    "print('stability of TF genes %s' % rate_df2.loc[rate_df2['TF'], 'stab_percentile'].median())\n",
    "#Percentile of non-CTS TF genes\n",
    "print('stability of non-CTS TF genes %s' % rate_df2.loc[rate_df2['TF'] & ~rate_df2['CTS'], 'stab_percentile'].median())\n",
    "#Percentile of TF & CTS\n",
    "print('stability of CTS & TF genes %s' % rate_df2.loc[rate_df2['TF'] & rate_df2['CTS'], 'stab_percentile'].median())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bacc0ef-e37b-4ceb-bd8e-82a756f932a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find CTS genes which are in the top and bottom 10% of most stable RNAs for the CTS genes\n",
    "# Previously I tried the top and bottom 10% of stability overall, but this found only 26 genes in the bottom 10% of genes\n",
    "# Use as bg set, all the genes with stability measurements\n",
    "bg_genes = rate_df2.index\n",
    "\n",
    "geneset_outdir = os.path.join(outdir, 'genesets')\n",
    "os.makedirs(geneset_outdir, exist_ok=True)\n",
    "\n",
    "num_genes = int(round(len(rate_df2.query('CTS'))/10))\n",
    "sorted_cts = rate_df2.query('CTS').sort_values(by='stab_percentile')\n",
    "stab_genes10 = sorted_cts.head(n=num_genes).index\n",
    "unstab_genes10 = sorted_cts.tail(n=num_genes).index\n",
    "\n",
    "# Write the output genelists\n",
    "pd.DataFrame(stab_genes10).to_csv(os.path.join(geneset_outdir, 'CTS_10unstable_genes.csv'), header=None, index=None)\n",
    "pd.DataFrame(unstab_genes10).to_csv(os.path.join(geneset_outdir, 'CTS_10stable_genes.csv'), header=None, index=None)\n",
    "pd.DataFrame(bg_genes).to_csv(os.path.join(geneset_outdir, 'bg_genes.csv'), header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7aa0652a",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrich_tf = pd.DataFrame(res_tf['enrich'], index=res_tf['order'], columns=['-log10_enrich_tf'])\n",
    "enrich_rbp = pd.DataFrame(res_rbp['enrich'], index=res_rbp['order'], columns=['-log10_enrich_rbp', '-log10_enrich_mrbp'])\n",
    "enrich_df = pd.concat([enrich_tf, enrich_rbp], axis=1)\n",
    "print(f'-log10 0.05 {-math.log(0.05, 10)}')\n",
    "enrich_df.sort_values(by='-log10_enrich_tf', ascending=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 ('pretty')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "bdb49103a4ed208a05ea4530afbe53462c06fafea10c7833b005d674746fdb08"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
